vcd/resource_vcd_vapp.go

```diff
@@ -4,24 +4,32 @@ import (
 	"fmt"
 	"log"
 	"regexp"
+	"strings"
 
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/vmware/go-vcloud-director/v2/govcd"
 	"github.com/vmware/go-vcloud-director/v2/types/v56"
 )
 
+const implicitVmInVappDeprecation = "Creation of vApp and an implicit single VM in the same structure is deprecated"
```

Comment:
There is an issue with providing a **generic** deprecation message to all the fields. In addition to the above message, we need to print "the new alternative" - i.e. what to use instead of the deprecated field. For example, if we really deprecate the `network_name` field (see concerns about it below), then we need to point the user to which parameter to use in the VM resource **specifically**.

All in all, all the deprecation messages should contain both the above **generic** statement + the **specific** statement on what to use instead.

CC: @Didainius @vbauzysvmware

Reply:
There is no replacement for the properties we are deprecating. The alternative is to define a VM explicitly.
Please suggest a wording that you think better conveys the meaning.

Reply:
Also, consider that deprecation works by showing a warning for every deprecated field. 
We are deprecating several fields, and so users that create a VM implicitly will see the same message half a dozen times. 
The alternative is to **deprecate the resource and make a new one with a different name.**

Reply:
> There is no replacement for the properties we are deprecating.

That's not true. Each of these fields have a counter part in the VM resources and we should point that out to the user.

> We are deprecating several fields, and so users that create a VM implicitly will see the same message half a dozen times.

That's one more reason for showing a customized (i.e. specific) message for each of the deprecated fields. This way the messages will be tailored.

Something like this:

```
network_name deprecated: In favor of vcd_vapp_vm.network. Creation of vApp and an implicit single VM in the same structure is deprecated.
```

This would also be in-line with what we have inside the VM resource itself:
```
			"network_name": &schema.Schema{
				ConflictsWith: []string{"network"},
				Deprecated:    "In favor of network",
				ForceNew:      true,
				Optional:      true,
				Type:          schema.TypeString,
			},
```

Reply:
I saw you made changes to the deprecation message could you show how the latest variant looks when running `terraform plan`? Thanks.

Reply:
Here's an example:
```
Terraform will perform the following actions:

  # vcd_network_routed.TestAccVcdVAppNet2 will be created
  + resource "vcd_network_routed" "TestAccVcdVAppNet2" {
      + dns1         = "8.8.8.8"
      + dns2         = "8.8.4.4"
      + edge_gateway = "gw-datacloud"
      + gateway      = "10.10.103.1"
      + href         = (known after apply)
      + id           = (known after apply)
      + name         = "TestAccVcdVAppNet2"
      + netmask      = "255.255.255.0"
      + org          = "datacloud"
      + shared       = false
      + vdc          = "vdc-datacloud"

      + dhcp_pool {
          + end_address    = "10.10.103.254"
          + max_lease_time = 7200
          + start_address  = "10.10.103.171"
        }

      + static_ip_pool {
          + end_address   = "10.10.103.170"
          + start_address = "10.10.103.2"
        }
    }

  # vcd_vapp.TestAccVcdVAppVappPowerOff will be created
  + resource "vcd_vapp" "TestAccVcdVAppVappPowerOff" {
      + accept_all_eulas = true
      + catalog_name     = "cat-datacloud"
      + cpus             = 1
      + href             = (known after apply)
      + id               = (known after apply)
      + ip               = "10.10.103.160"
      + memory           = 1024
      + metadata         = {
          + "vapp_metadata" = "vApp Metadata."
        }
      + name             = "TestAccVcdVAppVappPowerOff"
      + network_name     = "TestAccVcdVAppNet2"
      + org              = "datacloud"
      + power_on         = false
      + status           = (known after apply)
      + status_text      = (known after apply)
      + template_name    = "photon-hw11"
      + vdc              = "vdc-datacloud"
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Warning: "memory": [DEPRECATED] Creation of vApp and an implicit single VM in the same structure is deprecated. Use vcd_vapp_vm.memory instead

  on config.tf line 39, in resource "vcd_vapp" "TestAccVcdVAppVappPowerOff":
  39: resource "vcd_vapp" "TestAccVcdVAppVappPowerOff" {



Warning: "ip": [DEPRECATED] Creation of vApp and an implicit single VM in the same structure is deprecated. Use vcd_vapp_vm.network instead

  on config.tf line 39, in resource "vcd_vapp" "TestAccVcdVAppVappPowerOff":
  39: resource "vcd_vapp" "TestAccVcdVAppVappPowerOff" {



Warning: "template_name": [DEPRECATED] Creation of vApp and an implicit single VM in the same structure is deprecated. Use vcd_vapp_vm.template_name instead

  on config.tf line 39, in resource "vcd_vapp" "TestAccVcdVAppVappPowerOff":
  39: resource "vcd_vapp" "TestAccVcdVAppVappPowerOff" {



Warning: "catalog_name": [DEPRECATED] Creation of vApp and an implicit single VM in the same structure is deprecated. Use vcd_vapp_vm.catalog_name instead

  on config.tf line 39, in resource "vcd_vapp" "TestAccVcdVAppVappPowerOff":
  39: resource "vcd_vapp" "TestAccVcdVAppVappPowerOff" {



Warning: "network_name": [DEPRECATED] Creation of vApp and an implicit single VM in the same structure is deprecated. Use vcd_vapp_vm.network instead

  on config.tf line 39, in resource "vcd_vapp" "TestAccVcdVAppVappPowerOff":
  39: resource "vcd_vapp" "TestAccVcdVAppVappPowerOff" {



Warning: "cpus": [DEPRECATED] Creation of vApp and an implicit single VM in the same structure is deprecated. Use vcd_vapp_vm.cpus instead

  on config.tf line 39, in resource "vcd_vapp" "TestAccVcdVAppVappPowerOff":
  39: resource "vcd_vapp" "TestAccVcdVAppVappPowerOff" {


```

Reply:
That's actually perfect.